opflexagent/firewall_wrapper.py

```diff
@@ -0,0 +1,115 @@
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from oslo_log import log as logging
+from oslo_utils import importutils
+
+
+LOG = logging.getLogger(__name__)
+
+
+class FirewallDriverWrapper(object):
+    """Wrapper for Firewall Driver implemntations
+
+    This wrapper class keeps a map of drivers to use for
+    implementing a firewall driver. The firewall driver methods
+    are divided into two groups:
+      1) methods where only one driver implementation should be called
+      2) methods where all driver implementations should be called
+
+    The first group is called when there is only a single argument:
+    the neutron port. The second group is called in all other cases.
+    The first group uses a key to determine which driver gets invoked.
+    The key is obtained by a user-defined strategy function, which gets
+    passed the port object. The strategy function is meant to perform
+    some operation (e.g. search) on the port to determine and return
+    the key to use for looking up the appropriate firewall driver.
+
+    There is also a defaulit firewall driver. This is used when there
+    is only a single firewall driver is used, and therefore a key is
+    not needed.
+    """
+
+    def __init__(self):
+        self._strategy_map = {}
+        self._strategy_fn = None
+        self._default_strategy = None
+        self._port_methods = ['prepare_port_filter',
+                              'apply_port_filter',
+                              'update_port_filter',
+                              'remove_port_filter',
+                              'remove_port_filter']
+        self._all_strategy_methods = ['filter_defer_apply_on',
+                                      'filter_defer_apply_off',
+                                      'ports',
+                                      'defer_apply',
+                                      'update_security_group_members',
+                                      'update_security_group_rules']
+
+    def _do_port_strategy(self, name, port):
```

Comment:
signature here should be something like:

def _do_port_strategy(self, name, port, *args, **kwargs):